# Concrete Compressive Strength

## Importing Libraries

In [47]:
import keras
from keras.models import Sequential
from keras.layers import Dense, Input
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

### Importing Data and Preprocessing

In [48]:
concrete_data=pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')

In [49]:
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [50]:
concrete_data.shape

(1030, 9)

In [51]:
concrete_data.isna().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [52]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [53]:
concrete_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.6 KB


In [54]:
cols= concrete_data.columns

### Defining predictor and Target Variables

## Part A

In [56]:
predictorsX=concrete_data[cols[cols!='Strength']]

In [57]:
predictorsX.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [58]:
targetY=concrete_data['Strength']

In [59]:
targetY.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [60]:
pred_cols=predictorsX.shape[1]

### Defining the Model with one hidden layer

In [61]:
def regression_model1():
    
    model=Sequential()
    model.add(Input(shape=(pred_cols,)))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(1))
    
    model.compile(optimizer='adam',loss='mean_squared_error')
    
    return model

### Defining function to train model

In [62]:
def model_training(X,Y, regression_model,epochs):
    
    mse_list=[]
    
    for epoc in range(epochs):
    
        model=regression_model
        
        xtrain,xtest,ytrain,ytest=train_test_split(X,Y,test_size=0.3,random_state=42)
        
        model.fit(xtrain,ytrain,epochs=50,verbose=0)
        
        y_predict=model.predict(xtest)
        
        mse=mean_squared_error(ytest,y_predict)
        mse_list.append(mse)
        
        print(f"Run {epoc+1}, MSE: {mse}")
    
    # Calculate the mean and standard deviation of the MSE values
    mean_mse = np.mean(mse_list)
    std_mse = np.std(mse_list)
    
    print(f"Mean MSE: {mean_mse}")
    print(f"Standard Deviation of MSE: {std_mse}")
        

In [63]:
model_training(predictorsX,targetY,regression_model=regression_model1(),epochs=50)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 1, MSE: 230.71931653038166
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 2, MSE: 119.80531137125028
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Run 3, MSE: 93.28201999469883
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 4, MSE: 78.0637698310586
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 5, MSE: 63.470151061498555
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 6, MSE: 58.14891208560888
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 7, MSE: 51.559095975461034
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 8, MSE: 51.29153569282839
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 9, MSE: 50.968650576273866
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 10, MSE: 49.10793361575202
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 11, MSE: 49.235633341709594
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 12, MSE: 48.573175029803615
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 13, MSE: 58.87900500554641
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 14, MSE: 60.18527542489415
10/10 ━━━

### Normalising the data

## Part B

In [65]:
scaler=StandardScaler()

predictors=pd.DataFrame(scaler.fit_transform(predictorsX),columns=predictorsX.columns)
target = pd.DataFrame(scaler.fit_transform(targetY.values.reshape(-1, 1)), columns=['target'])

predictors.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
1  2.477915           -0.856888 -0.847144 -0.916764         -0.620448   
2  0.491425            0.795526 -0.847144  2.175461         -1.039143   
3  0.491425            0.795526 -0.847144  2.175461         -1.039143   
4 -0.790459            0.678408 -0.847144  0.488793         -1.039143   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.863154       -1.217670 -0.279733  
1          1.056164       -1.217670 -0.279733  
2         -0.526517       -2.240917  3.553066  
3         -0.526517       -2.240917  5.057677  
4          0.070527        0.647884  4.978487

In [66]:
target.head()

target
0  2.645408
1  1.561421
2  0.266627
3  0.313340
4  0.507979

In [67]:
model_training(predictors,target,regression_model=regression_model1(),epochs=50)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 1, MSE: 0.23058902216231733
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Run 2, MSE: 0.16388995623054925
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 3, MSE: 0.152897291004724
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 4, MSE: 0.15018084099400736
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 5, MSE: 0.14665957969817725
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 6, MSE: 0.14493377408814292
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 7, MSE: 0.14374800171005864
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 8, MSE: 0.14017069860083034
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 9, MSE: 0.13526830161477543
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Run 10, MSE: 0.13395114759394408
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 11, MSE: 0.12937271862000627
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 12, MSE: 0.12510341601588593
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 13, MSE: 0.12445463526106948
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 14, MSE: 0.1224752

The mean squared error (MSE) of the model decreased dramatically from 57.15 in Part A to 0.128 in Part B, despite using the same number of epochs for training.

Insights:
Effect of Data Normalization: The key difference between Part A and Part B is the normalization of the data. Normalizing the predictors likely improved the model's ability to converge by ensuring that all features contributed proportionally to the learning process. This step helps prevent large-scale features from dominating smaller-scale ones, leading to a significant reduction in prediction error.

Same Epochs, Better Results: The same number of epochs achieved better performance in Part B due to improved input data quality, emphasizing the importance of preprocessing in deep learning.

## Part C

In [69]:
model_training(predictors,target,regression_model=regression_model1(),epochs=100)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
Run 1, MSE: 0.20876835680507777
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 
Run 2, MSE: 0.16963520845966798
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step
Run 3, MSE: 0.1626825832057825
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 4, MSE: 0.15766278776029052
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step  
Run 5, MSE: 0.15002866183207866
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step  
Run 6, MSE: 0.1457243912065982
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step 
Run 7, MSE: 0.1457144962452833
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 8, MSE: 0.14183553237746294
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 9, MSE: 0.13947058703526505
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 10, MSE: 0.1394210246280596
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 11, MSE: 0.13801589449794174
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 12, MSE: 0.13642319126903693
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Run 13, MSE: 0.13665119754550534
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 14, MSE: 0.1350366

The mean squared error (MSE) increased slightly from 0.128 in Part B to 0.138 in Part C, despite increasing the number of epochs during training.

Insights:
Diminishing Returns on Epochs: The slight increase in MSE indicates that simply increasing the number of epochs does not always lead to better performance. This could be due to the model reaching its optimal learning capacity within the initial epochs, and further training may introduce noise or overfitting.

Plateau in Model Performance: The lack of significant improvement suggests that the model might have already converged in Part B. Additional epochs are not contributing meaningful learning and could instead be causing minor fluctuations in performance.

### Defining model with 3 hidden layers

## Part D

In [71]:
def regression_model2():

    model=Sequential()
    model.add(Input(shape=(pred_cols,)))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(1))
    
    model.compile(optimizer='adam',loss='mean_squared_error')
    
    return model

In [83]:
model_training(predictors,target,regression_model=regression_model2(),epochs=50)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
Run 1, MSE: 0.18462158980362564
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
Run 2, MSE: 0.16048986508870816
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 3, MSE: 0.15759415604827692
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 4, MSE: 0.15404935223153002
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 5, MSE: 0.14958466949899288
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 6, MSE: 0.14645642582019525
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 7, MSE: 0.1442186207645002
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 8, MSE: 0.14259625392652212
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 9, MSE: 0.14143552357397376
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
Run 10, MSE: 0.13830933424130964
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 11, MSE: 0.13927915567275645
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 12, MSE: 0.1373793233713859
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 
Run 13, MSE: 0.13512744530635512
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
Run 14, MSE: 0.1341702

The mean squared error (MSE) increased from 0.128 in Part B to 0.138 in Part D when the number of hidden layers was increased.

Insights:
Impact of Additional Complexity: Increasing the number of hidden layers adds complexity to the model, potentially making it harder to generalize to unseen data. This could explain the slight increase in MSE, as the model may have started to overfit the training data or faced challenges in optimizing effectively.

Saturation of Learning: The performance plateau suggests that the dataset and problem complexity may not require additional hidden layers. For relatively simple tasks or datasets, adding more layers can introduce redundancy and unnecessary computational overhead without improving results.

In [84]:
def regression_model():
    
    model=Sequential()
    model.add(Input(shape=(pred_cols,)))
    model.add(Dense(100,activation='relu'))
    model.add(Dense(100,activation='relu'))
    model.add(Dense(1))

    model.compile(optimizer='adam',loss='mean_squared_error')

    return model

In [85]:
model=regression_model()
print(model)

<Sequential name=sequential_207, built=True>


In [86]:
model_training(predictors,target,regression_model=regression_model(),epochs=50)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step
Run 1, MSE: 0.10994450366759112
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
Run 2, MSE: 0.09215148951827913
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 3, MSE: 0.10415949157450806
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
Run 4, MSE: 0.1006872824279386
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 5, MSE: 0.09930534807903013
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 6, MSE: 0.09606451478991945
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 7, MSE: 0.09993776524167408
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 8, MSE: 0.09387736092936191
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
Run 9, MSE: 0.10590069836506959
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step 
Run 10, MSE: 0.09409904332017477
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
Run 11, MSE: 0.09806748282744616
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 12, MSE: 0.09415499166925512
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 13, MSE: 0.09935062591805348
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
Run 14, MSE: 0.09529

In [87]:
model.fit(predictors,targetY, validation_split=0.3,epochs=100, verbose=2)

Epoch 1/100
23/23 - 13s - 549ms/step - loss: 1588.0471 - val_loss: 1053.3977
Epoch 2/100
23/23 - 0s - 20ms/step - loss: 1269.5583 - val_loss: 777.5127
Epoch 3/100
23/23 - 1s - 30ms/step - loss: 806.3644 - val_loss: 433.1950
Epoch 4/100
23/23 - 0s - 19ms/step - loss: 386.2609 - val_loss: 214.1478
Epoch 5/100
23/23 - 0s - 19ms/step - loss: 251.1802 - val_loss: 175.8321
Epoch 6/100
23/23 - 0s - 20ms/step - loss: 226.3924 - val_loss: 169.0334
Epoch 7/100
23/23 - 0s - 20ms/step - loss: 207.0029 - val_loss: 163.3777
Epoch 8/100
23/23 - 1s - 28ms/step - loss: 195.1004 - val_loss: 158.9497
Epoch 9/100
23/23 - 0s - 22ms/step - loss: 185.4662 - val_loss: 156.2041
Epoch 10/100
23/23 - 1s - 24ms/step - loss: 177.6386 - val_loss: 153.8455
Epoch 11/100
23/23 - 1s - 25ms/step - loss: 170.8777 - val_loss: 154.2149
Epoch 12/100
23/23 - 1s - 25ms/step - loss: 166.5253 - val_loss: 153.8207
Epoch 13/100
23/23 - 0s - 19ms/step - loss: 161.3666 - val_loss: 149.3184
Epoch 14/100
23/23 - 0s - 21ms/step - loss

In [108]:
scores = model.evaluate(predictors, targetY, verbose=2)

33/33 - 0s - 12ms/step - loss: 60.9356


In [110]:
print('Accuracy: {:.2f}% \nError: {:.2f}'.format(scores, (1 - scores)))


Accuracy: 60.94% 
Error: -59.94


In [90]:
model.save('regression_model.keras')